README.md
<h1 align="center">
  <img alt="IDENA Runner Bash Script - fast idena-go network node deployment with possibility of multiple instances installation: 1 user - 1 idena-go node instance" src="https://github.com/ltraveler/ltraveler/raw/main/images/idena-runner-logo.png" width="224px"/><br/>
  Idena RUNNER Script
</h1>
<p align="center"><b>Bash Script implementation</b> of the <b>Idena network node</b> installation wizard. <br> Install multiple instances of the <b>Idena-Go</b> in a simple and user-friendly way.</p>

<p align="center"><a href="https://github.com/ltraveler/idena-runner/releases/latest" target="_blank"><img src="https://img.shields.io/github/v/release/ltraveler/idena-runner?style=for-the-badge&logo=none" alt="idena runner latest version" /></a>&nbsp;<a href="https://wiki.ubuntu.com/FocalFossa/ReleaseNotes" target="_blank"><img src="https://img.shields.io/badge/Ubuntu-20.04(LTS)+-00ADD8?style=for-the-badge&logo=none" alt="Ubuntu minimum version" /></a>&nbsp;<a href="https://github.com/ltraveler/idena-runner/blob/main/CHANGELOG.md" target="_blank"><img src="https://img.shields.io/badge/Build-Stable-success?style=for-the-badge&logo=none" alt="idena-go latest release" /></a>&nbsp;<a href="https://www.gnu.org/licenses/quick-guide-gplv3.html" target="_blank"><img src="https://img.shields.io/badge/license-GPL3.0-red?style=for-the-badge&logo=none" alt="license" /></a>&nbsp;<a href="https://github.com/ltraveler/idena-runner/blob/main/README.ru-RU.md" target="_blank"><img src="https://img.shields.io/badge/readme-РУССКИЙ-orange?style=for-the-badge&logo=none" alt="Скрипт Idena Runner" /></a></p>

## 🚀&nbsp; Running `idena_install.sh` (requires root privileges)

Please make sure that you have a pure Ubuntu 20.04 installation.
To install Idena node using this script, please folow these steps:
* `apt-get install -y git` installing git package
* `git clone https://github.com/ltraveler/idena-runner.git` clone the repository
* `cd idena-runner`
* `chmod +x idena_install.sh` to make the script executable
* `./idena_install.sh` to run the script

## ✅&nbsp; Features

* Multiple Idena instances installation: 1 user - 1 instance
* Import the existing private/node keys during the installation process
* Automatic updates crontask that can be schedulled during the installation process
* Uncomplicated Firewall (UFW) configuration and automatic port rules updates during the idena-node instance installaltion
* Possibility to install idena-go as a shared node with default recommended values

## 🙋&nbsp; What the script is doing?

1. Checking if the idena.service exists;
2. Creating new user and password to run the Idena node daemon;
3. Upgrading Ubuntu packages and installing all requiered dependecies;
4. Downloading idena-go network node based on the version that user have entered. If the input is empty the script is downloading the latest one. The version history is available [here](https://github.com/idena-network/idena-go/releases);
5. The script using pre-defined `config.json` file which can be changed during the installation process;
6. Installing Idena-go and running it based on the config.json file from the repository;
7. Changing API and Private keys of the node to the custom ones if the user wants so;
8. Creating cron job to check for idena-go updates ones once a day. You can specify the frequency during the installation process;
9. Creating Idena Daemon and running it;
10. Installing and running firewall based on SSH and IPFS port numbers.

### 🏛️&nbsp;  In case if you are installing a shared node:
1. The script will add `--profile=shared` to the service file;
2. You can set the most important args: `BlockPinThreshold`, `FlipPinThreshold`, `AllFlipsLoadingTime`
   - **Default values:** 
     - `BlockPinThreshold` = `0.3`
     - `FlipPinThreshold` = `1`
     - `AllFlipsLoadingTime` = `7200000000000`

##  ⚙️&nbsp;  About Idena Daemon
The script is creating a service daemon called idena. Which starts on the boot.
#### You can use these commands to control it:
* `service idena_$username status`- to check the status 
* `service idena_$username restart` - to restart the service
* `service idena_$username stop` - to stop the service
* `service idena_$username start` - to start the service

*where $username is required instance username

##  💻&nbsp;  Command Line Flags and Arguments
Since version _0.3.0_ the script could be run in silence mode (aka **push installation**).
All or part of the answers could be sent via flags and relevant arguments in the command line.
**flags:**\
            `-u` or `--username` - _username_\
            `-p` or `--password` - _password_ in case of using `-u` without `-p` the password would be the same as username\
            `-s` or `--shared` - _shared node installation_\
            `-v` or `--version` - _idena-go node client version_ or _latest_ to download the latest one\
            `-b` or `--blockpinthreshold` - _Block Pin Threshold_ if not set but `-f` and/or `-l` have been applied, the script will use the default recommended value [`0.3`]\
            `-f` or `--flippinthreshold` - _Flip Pin Threshold_ if not set but `-b` and/or `-l` have been applied, the script will use the default recommended value [`1`]\
            `-l` or `--allflipsloadingtime` - _All Flips Loading Time_ if not set but `-b` and/or `-f` have been applied, the script will use the default recommended value [`7200000000000`]\
            `-r` or `--rpcport` - _RPC Port_ aka _HTTP Port_\
            `-i` or `--ipfsport` - _IPFS Port_\
            `-k` or `--privatekey` - _IDENA Private Key_ aka _nodekey_\
            `-a` or `--apikey` - _IDENA Node API Key_\
            `-d` or `--updatefreq` - _Update frequency in CRON expression format_

**Apart from `-s` or `--shared` the rest of the flags need an argument inside '' (_apostrophe_)**

_**For example:**_\
`./idena_install.sh -u 'ratel' -p 'ratel' -s -v 'latest' -b '0.3' -f '1' -l '7200000000000' -r '9189' -i '41283' -k '6e17f7490f7922f3224d41769ce5ed2a01030de69d77163a291a77e8280aad3' -a '33e32cd86ecfb3179e50208428541a1e' -d '0 0 * * *'`

## ✔️&nbsp; Idena-runner instance update process (requires root privileges)

1. Backup the private key (`/home/idena_instance_username/idena-go/datadir/keystore/nodekey`)
2. Backup the node api.key (`/home/idena_instance_username/idena-go/datadir/api.key`)
3. Run the latest version of the idena-runner script and put the same username that you have used to install the instance that you are trying to update.
That will overwrite all required files.
***Attention:*** **all files inside the idena-go folder will be permanently deleted**.
4. `service idena_$username stop` - to stop the updated instance.
5. Restore your private and API keys from the backup that you have made on the 1st step.
6. `service idena_$username start` - to start the updated instance.

## 🗑️&nbsp; Idena-go instance uninstallation process (requires root privileges)

1. `service idena_username stop` _stopping idena user instance_;
2. `pkill -u username` _killing all processes related to the user_;
3. `deluser --remove-home username` _removing related to idena-go instance user and all his files and folders_;
4. `rm /etc/cron.d/idena_update_username` _removing cron idena-go update related task_;
5. `rm /etc/systemd/system/idena_username.service` _removing idena daemon service related to the instance that we are uninstalling_;
6. `systemctl disable idena_username` _disabling idena daemon service related to the instance that we are uninstalling_
7. `systemctl daemon-reload` and `systemctl reset-failed` _updating systemctl changes that we have made in the previous step_;
8. `ufw show added | grep "IDENA Instance for user $username" | awk '{ gsub("ufw","ufw delete",$0); system($0)}'` do not forget to replace `$username` to the real name of deleted user;
9. `sudo visudo` you have to find and delete the line related to the deleted user at the end of the file;
10. `nano /etc/ssh/sshd_config` - you have to delete username from _DenyUsers_ group;
11. `sudo sshd -t` - to be sure that your latest changes of _sshd_config_ file is correct;
12. `sudo service sshd restart` - to restart _sshd daemon_.

### 🤝&nbsp; Idena Donations


CHANGELOG.md
# Changelog

## 0.4.3 (Nov 28, 2022)

### Bug Fixes

- SSH port detection fix

## 0.4.2 (Nov 27, 2022)

### Bug Fixes

- **ARM64**: update script won't start upgrading process in case if `idena-go` binary has the same size as the local one

## 0.4.1 (Nov 26, 2022)

### Changes

- Now Idena Runner compatible with `ARM64` servers like `Raspberry Pi`, `Ampere A1 Compute` and so on.<br><b>ATTENTION:</b> <ins><i>If you have a mobile device you have to use</i></ins> <b>[IDENA ARMer script](https://github.com/ltraveler/idena-armer)</b>.


## 0.3.7 (July 8, 2022)

### Bug Fixes

- Wrong call to Idena node upgrade script

## 0.3.6 (June 28, 2022)

### Bug Fixes

- In the case of importing private and node keys manually, the import process kept the default generated files instead of writing new values

## 0.3.5 (Apr 07, 2022)

- The password will be set the same as username in case if there is an empty answer on the password prompt in interactive mode 

## 0.3.4 (Mar 31, 2022)

### Bug Fixes

- ufw rules management updates
- correction related to existed screen session precheck

## 0.3.3 (Mar 31, 2022)

### Bug Fixes

- `--username` long option didn't work properly

## 0.3.2 (Mar 31, 2022)

### Bug Fixes

- bug fix related to use --long-options and corresponding arguments

## 0.3.1 (Mar 30, 2022)

### Bug Fixes

- some critical bugs have been fixed.\
**Attention:** _**v0.3.0 must be upgraded to the latest one**_

## 0.3.0 (Mar 30, 2022)

### Changes

- command line flags model has been completely reorganized.\
**flags:**\
            `-u` or `--username` - _username_\
            `-p` or `--password` - _password_ in case of using `-u` without `-p` the password would be the same as username\
            `-s` or `--shared` - _shared node installation_\
            `-v` or `--version` - _idena-go node client version_ or _latest_ to download the latest one\
            `-b` or `--blockpinthreshold` - _Block Pin Threshold_ if not set but `-f` and/or `-l` have been applied, the script will use the default recommended value [`0.3`]\
            `-f` or `--flippinthreshold` - _Flip Pin Threshold_ if not set but `-b` and/or `-l` have been applied, the script will use the default recommended value [`1`]\
            `-l` or `--allflipsloadingtime` - _All Flips Loading Time_ if not set but `-b` and/or `-f` have been applied, the script will use the default recommended value [`7200000000000`]\
            `-r` or `--rpcport` - _RPC Port_ aka _HTTP Port_\
            `-i` or `--ipfsport` - _IPFS Port_\
            `-k` or `--privatekey` - _IDENA Private Key_ aka _nodekey_\
            `-a` or `--apikey` - _IDENA Node API Key_\
            `-d` or `--updatefreq` - _Update frequency in CRON expression format_

**Apart from `-s` or `--shared` the rest of the flags need an argument inside '' (_apostrophe_)**

_**For example:**_\
`./idena_install.sh -u ratel -p ratel -s -v 'latest' -b '0.3' -f '1' -l '7200000000000' -r '9189' -i '41283' -k '6e17f7490f7922f3224d41769ce5ed2a01030de69d77163a291a77e8280aad3' -a '33e32cd86ecfb3179e50208428541a1e' -d '0 0 * * *'`

## 0.2.9 (Mar 24, 2022)

### Changes

- flags `-u` and `-p` have been added as possible arguments to run the script 

## 0.2.8 (Mar 17, 2022)

### Changes

- Some improvements in user experience model

## 0.2.7 (Mar 15, 2022)

### Changes

- Some color style changes for better user experience

### Bug Fixes

- Small bug fixes related to SSH deny permissions for re-created usernames

## 0.2.6 (Mar 06, 2022)

### Changes

- Some changes in user-interactive dialog to make it more user-friendly
- Possibility to set all default values in `config.json` in case of the shared node installation

## 0.2.5 (Feb 09, 2022)

### Changes

- During the shared node installation wizard user can modify some important args: `BlockPinThreshold`, `FlipPinThreshold`, `AllFlipsLoadingTime`
- Now the script is overwriting `config.json` file to the default one to prevent possible conflicts in case of multiple installations

### Bug Fixes


idena_install.sh
#!/bin/bash 
#
#ASCII color assign
RED="\033[0;31m"
LRED="\033[1;31m"
YELLOW="\033[0;33m"
LYELLOW="\033[1;33m"
GREEN="\033[0;32m"
LGREEN="\033[1;32m"
BLUE="\033[0;34m"
LBLUE="\033[1;34m"
PURPLE="\033[0;35m"
LPURPLE="\033[1;35m"
CYAN="\033[0;36m"
LCYAN="\033[1;36m"
NC="\033[0m" # No Color
#input flags set
set -o pipefail -o nounset -o noclobber  #-o errexit 

! getopt --test > /dev/null 
if [[ ${PIPESTATUS[0]} -ne 4 ]]; then
    echo 'I’m sorry, `getopt --test` failed in this environment.'
    exit 1
fi

OPTIONS=u:p:sv:b:f:l:r:i:k:a:d: 
LONGOPTS=username:,password:,shared,version:,blockpinthreshold:,flippinthreshold:,allflipsloadingtime:,rpcport:,ipfsport:,privatekey:,apikey:,updatefreq:

! PARSED=$(getopt --options=$OPTIONS --longoptions=$LONGOPTS --name "$0" -- "$@")
if [[ ${PIPESTATUS[0]} -ne 0 ]]; then
    exit 2
fi
# 
eval set -- "$PARSED"
username='n' password='n' version='n' shared_node='n' bp_threshold='0.3' fp_threshold='1' af_time='7200000000000' nrpcport='n' nipfsport='n' privatekey='n' nodeapikey='n' idupdate='n' 
if [ "$shared_node" = "true" ]; then
echo -e 'incorrect argument for ${LYELLOW}-s${NC} | ${LYELLOW}--shared ${NC}flag'
fi
# 
while true; do
    case "$1" in
        -u|--username)
            username="$2"
            shift 2
            ;;
        -p|--password)
            password="$2"
            shift 2
            ;;
        -s|--shared)
            shared_node=yes
            shift
            ;;
        -v|--version)
            version="$2"
            shift 2
            ;;
        -b|--blockpinthreshold)
            bp_threshold="$2"
            shift 2
            ;;
        -f|--flippinthreshold)
            fp_threshold="$2"
            shift 2
            ;;
        -l|--allflipsloadingtime)
            af_time="$2"
            shift 2
            ;;
        -r|--rpcport)
            nrpcport="$2"
            shift 2
            ;;
        -i|--ipfsport)
            nipfsport="$2"
            shift 2
            ;;
        -k|--privatekey)
            privatekey="$2"
            shift 2
            ;;
        -a|--apikey)
            nodeapikey="$2"
            shift 2
            ;;
        -d|--updatefreq)
            idupdate="$2"
            shift 2
            ;;
        --)
            shift
            break
            ;;
        *)
            echo "Programming error"
            exit 3
            ;;
    esac
done
#password the same as username if password not specified 
if [[ "$username" != 'n' && -z "$password" ]]; then
password="$username"
fi
#updating Ubuntu and installing all required dependencies
apt-get update
apt-get upgrade -y
reqpkgs=('jq' 'git' 'ufw' 'curl' 'wget' 'nano' 'screen' 'psmisc' 'unzip' 'netcat-traditional')
for x in "${reqpkgs[@]}"; do
dpkg -s "$x" &> /dev/null
if [ $? != 0 ]; then
    echo -e "${LRED}Package $x is not instlled. Installing...${NC}"
    apt-get install -y $x; 
#
fi
done
#
echo -e "${LYELLOW}Please enter a ${LRED}username${NC} and ${LRED}password${LYELLOW} that you would like to use for this ${LGREEN}Idena Node Daemon${LYELLOW} instance.${NC}"
# read -p "Enter username : " username
if [ "$username" = 'n' ]; then
while
echo -e "${NC}Please do not use ${LGREEN}root${NC} as a username"
read -p "Enter username : " username
[[ $username = 'root' ]]
do true; done

read -s -p "Enter password : " password
if [ -z "$password" ]; then
password="$username"
fi
fi
#adding user to DenyUsers group
a_users=$(grep 'DenyUsers' /etc/ssh/sshd_config) 
if [ -z "$a_users" ]; then
    echo "User is not in the DenyUsers group"
    echo "User has been added to DenyUsers group"
    echo "DenyUsers $username" >> /etc/ssh/sshd_config
elif
    echo "$a_users" | grep -w "$username"; then
    echo "User already in DenyUsers list"
else
    a_users="$a_users $username"
    sed -i.bak "/DenyUsers/c$a_users" /etc/ssh/sshd_config
    echo "User has been added to DenyUsers group"
fi
service ssh restart
#Overwriting config.json to the default one from the repository to prevent possible conflicts
curl https://raw.githubusercontent.com/ltraveler/idena-runner/main/config.json >| config.json
#Are we installing a shared node?
if [ "$shared_node" = 'n' ]; then
while true; do
    read -p "$(echo -e ${LYELLOW}Would you like to install your idena node as a shared node? ${LGREEN}[y/N]${NC} )" yn
    case $yn in
        [Yy]* ) shared_node="true" && sed -i '/ExecStart/cExecStart=\/home\/$username\/idena-go\/idena-node --config=\/home\/$username\/idena-go\/config.json --profile=shared' idena.service && echo "Shared node installation"; break;; 
        [Nn]* ) sed -i 's/ --profile=shared//g' idena.service && echo "Regular node installation"; break;;
        * ) echo "Please answer yes or no.";;
    esac
done
else
    sed -i '/ExecStart/cExecStart=\/home\/$username\/idena-go\/idena-node --config=\/home\/$username\/idena-go\/config.json --profile=shared' idena.service && echo "Shared node installation" 
    sed -i -e "/BlockPinThreshold/c\    \"BlockPinThreshold\": $bp_threshold," -e "/FlipPinThreshold/c\    \"FlipPinThreshold\": $fp_threshold" -e "/AllFlipsLoadingTime/c\    \"AllFlipsLoadingTime\": $af_time," config.json; 
fi
#If shared true 
if [ "$shared_node" = "true" ]; then
while true; do
    read -p "$(echo -e ${LYELLOW}Would you like to set ${LRED}default recommended values${LYELLOW} in the shared node configuration file? ${LGREEN}[y/N]${NC} )" yn
    case $yn in
        [Yy]* ) echo -e "${LGREEN}Let's optimize shared node configuration file ${LYELLOW}by setting default values of\n${LRED}BlockPinThreshold: ${GREEN}-0.3${NC};\n${LRED}FlipPinThreshold: ${LGREEN}-1${NC};\n${LRED}AllFlipsLoadingTime: ${GREEN}-7200000000000${NC};" && bp_threshold=${bp_threshold:-0.3} && fp_threshold=${fp_threshold:-1} && af_time=${af_time:-7200000000000} && sed -i -e "/BlockPinThreshold/c\    \"BlockPinThreshold\": $bp_threshold," -e "/FlipPinThreshold/c\    \"FlipPinThreshold\": $fp_threshold" -e "/AllFlipsLoadingTime/c\    \"AllFlipsLoadingTime\": $af_time," config.json; break;;
        [Nn]* ) echo -e "${LGREEN}Let's optimize our shared node configuration file.${NC}\n${LRED}If you don't know the meaning of the following args, please skip them by pressing ENTER.${NC}\n${LGREEN}That will set the value to the default recommended one. You can see it inside the square brackets.${NC}" && read -p "$(echo -e ${LRED}BlockPinThreshold ${NC}[${LGREEN}0.3${NC}]: )" bp_threshold && bp_threshold=${bp_threshold:-0.3} && read -p "$(echo -e ${LRED}FlipPinThreshold ${NC}[${LGREEN}1${NC}]: )" fp_threshold && fp_threshold=${fp_threshold:-1} && read -p "$(echo -e ${LRED}AllFlipsLoadingTime ${NC}[${LGREEN}7200000000000${NC}]: )" af_time && af_time=${af_time:-7200000000000} && echo $bp_threshold && echo $fp_threshold && echo $af_time && sed -i -e "/BlockPinThreshold/c\    \"BlockPinThreshold\": $bp_threshold," -e "/FlipPinThreshold/c\    \"FlipPinThreshold\": $fp_threshold" -e "/AllFlipsLoadingTime/c\    \"AllFlipsLoadingTime\": $af_time," config.json; break;;
        * ) echo -e "${GREEN}Please answer yes or no.${NC}";;
    esac
done
fi
#checking if there is any idena daemon related to the inserted user
if [ -f "/etc/systemd/system/idena_$username.service" ]
then
echo "The service is exists. Don't worry, I will kill it."
systemctl stop idena_$username
systemctl disable idena_$username
ipfsport=($(jq -r '.IpfsConf.IpfsPort' /home/$username/idena-go/config.json))
ufw delete allow ${ipfsport[0]}
rm /etc/systemd/system/idena_$username.service ||: #killing the service
rm /usr/lib/systemd/system/idena_$username.service ||: # and related symlinks
systemctl daemon-reload
systemctl reset-failed
else
   echo "Making clean IDENA Node Installation"
fi
#Firewall precheck
ufw disable
if compgen -G "/etc/systemd/system/idena*.service" > /dev/null; then
echo "Idena Service exists. I will keep UFW rules as it is."
else
    echo "Applying UFW deny incoming rules for the first idena node instance installation"
    ufw default deny incoming
fi
#in case if the user has been deleted and screen session still exists
if pgrep screen &> /dev/null ; then sudo killall screen ; fi
#
# creating a user name and password for idena service 
#
if [ "$(id -u)" -eq 0 ]; then
	egrep -w "$username" /etc/passwd > /dev/null
	if [ $? -eq 0 ]; then
		echo "$username exists! Let's kill existed idena installation."
        rm -r /home/$username/idena-go ||:
        rm -f /etc/cron.d/idena_update_$username ||:
        pass=$(openssl passwd -5 $password)
        usermod --password $pass $username
        grep -q "$username ALL=NOPASSWD: /usr/sbin/service idena_$username *" /etc/sudoers || echo "$username ALL=NOPASSWD: /usr/sbin/service idena_$username *" >> /etc/sudoers
	else
        pass=$(openssl passwd -5 $password)
		useradd -s /bin/bash -m -p $pass $username
		[ $? -eq 0 ] && echo "User has been added to system!" || echo "Failed to add a user!"
        #add only if sudoers record doesn't exist 
        grep -q "$username ALL=NOPASSWD: /usr/sbin/service idena_$username *" /etc/sudoers || echo "$username ALL=NOPASSWD: /usr/sbin/service idena_$username *" >> /etc/sudoers 
	fi
else
	echo "Only root may add a user to the system"
	exit 2
fi
#
chown $username:$username /run/screen/S-$username
mkdir /home/$username/idena-go
#cd /home/$username/idena-go  
#downloading specific version or the latest one
if [[ $(dpkg --print-architecture) == *arm64* ]]; then version="arm64"; fi
if [ "$version" = 'latest'  ]; then version=$(curl -s https://api.github.com/repos/idena-network/idena-go/releases/latest | grep -Po '"tag_name":.*?[^\\]",' | sed 's/"tag_name": "v//g' |  sed 's/",//g') ; echo Installing version $version; fi 
if [ "$version" = 'n'  ]; then read -p "$(echo -e ${LYELLOW}Enter the number of the idena-go version \(eg. ${LRED}0.18.2\)${LYELLOW} ${LGREEN}keep it empty ${LYELLOW}to download the latest one:${NC} )" version; fi
if [ -z $version ]; then version=$(curl -s https://api.github.com/repos/idena-network/idena-go/releases/latest | grep -Po '"tag_name":.*?[^\\]",' | sed 's/"tag_name": "v//g' |  sed 's/",//g') ; echo Installing version $version; fi
touch /home/$username/idena-go/version
echo "$version" >| /home/$username/idena-go/version
if [ "$version" != 'arm64' ]; then wget https://github.com/idena-network/idena-go/releases/download/v$version/idena-node-linux-$version; else wget https://github.com/ltraveler/idena-go-arm64/raw/main/idena-go; fi
#customize config.json

if [[ "$nrpcport" = 'n' && "$nipfsport" != 'n' ]]; then
    read -p "$(echo -e ${LRED}HTTP Port ${LYELLOW}aka ${LRED}RPC Port ${NC}[${LGREEN}9009${NC}]: )" nrpcport && nrpcport=${nrpcport:-9009}
    sed -i -e "/HTTPPort/c\    \"HTTPPort\": $nrpcport" -e "/IpfsPort/c\    \"IpfsPort\": $nipfsport," config.json
elif [[ "$nipfsport" = 'n' && "$nrpcport" != 'n' ]]; then
    read -p "$(echo -e ${LRED}IPFS Port ${NC}[${LGREEN}40405${NC}]: )" nipfsport && nipfsport=${nipfsport:-40405}
    sed -i -e "/IpfsPort/c\    \"IpfsPort\": $nipfsport," -e "/HTTPPort/c\    \"HTTPPort\": $nrpcport" config.json
elif [[ "$nipfsport" != 'n' && "$nrpcport" != 'n' ]]; then
    sed -i -e "/HTTPPort/c\    \"HTTPPort\": $nrpcport" -e "/IpfsPort/c\    \"IpfsPort\": $nipfsport," config.json
else
while true; do
    read -p "$(echo -e ${LYELLOW}If you are installing multiple instances of Idena Node, ${LRED}you have to change default ports in ${LGREEN}config.json ${LRED}file. ${LYELLOW}Would you like to do so? ${LGREEN}[y/N]${NC})" yn
    case $yn in
        [Yy]* ) nano config.json; break;; 
        [Nn]* ) echo "Using default config.json file"; break;;
        * ) echo "Please answer yes or no.";;
    esac
done
fi
    
#copying config.json and idena.service
cp {config.json,idena.service} /home/$username/idena-go
if [ "$version" != 'arm64' ]; then mv idena-node-linux-$version /home/$username/idena-go/idena-node; else mv idena-go /home/$username/idena-go/idena-node; fi
chown -R $username:$username /home/$username/idena-go
#checking if ipfs port is opened
ipfsport=($(jq -r '.IpfsConf.IpfsPort' /home/$username/idena-go/config.json))
until ! nc -zv localhost $ipfsport; do
echo IPFS port is already used. Please choose another one.; read -p "Press enter to edit config.json file"; nano /home/$username/idena-go/config.json; 
ipfsport=($(jq -r '.IpfsConf.IpfsPort' /home/$username/idena-go/config.json))
done
#checking if rpc port is opened
rpcport=($(jq -r '.RPC.HTTPPort' /home/$username/idena-go/config.json))
until ! nc -zv localhost $rpcport; do
echo RPC port is already used. Please choose another one.; read -p "Press enter to edit config.json file"; nano /home/$username/idena-go/config.json;
rpcport=($(jq -r '.RPC.HTTPPort' /home/$username/idena-go/config.json))
done
#iDNA blockchain bootstrp mirrors
dnabc="https://sync.idena.site/idenachain.db.zip"
dnabc1="https://sync.idena-ar.com/idenachain.db.zip"
dnabc2="https://github.com/ltraveler/idenachain.db.git"
function validate_url()
{
    wget --spider $1
    return $?
}
mkdir /home/$username/idena-go/datadir ; mkdir /home/$username/idena-go/datadir/idenachain.db
chown -R $username:$username /home/$username/idena-go/datadir
if validate_url $dnabc2; then echo -e "${YELLOW}Downloading IDENA blockchain bootstrap: GitHub${NC}" ; rm -rf /home/$username/idena-go/datadir/idenachain.db ; git clone -b main --depth 1 --single-branch $dnabc2 /home/$username/idena-go/datadir/idenachain.db && rm -rf /home/$username/idena-go/datadir/idenachain.db/.git; elif validate_url $dnabc; then echo -e "${YELLOW}Downloading IDENA blockchain bootstrap: Mirror 01${NC}" &&  rm -rf /home/$username/idena-go/datadir/idenachain.db && wget --directory-prefix=/home/$username/idena-go/datadir $dnabc && unzip /home/$username/idena-go/datadir/idenachain.db.zip -d /home/$username/idena-go/datadir && rm -f /home/$username/idena-go/datadir/idenachain.db.zip; elif validate_url $dnabc1; then echo -e "${YELLOW}Downloading IDENA blockchain bootstrap: Mirror 2${NC}" && rm -rf /home/$username/idena-go/datadir/idenachain.db &&  wget --directory-prefix=/home/$username/idena-go/datadir $dnabc1 && unzip /home/$username/idena-go/datadir/idenachain.db.zip -d /home/$username/idena-go/datadir && rm -f /home/$username/idena-go/datadir/idenachain.db.zip; else echo "IDENA blockchain mirror is not available"; fi;
#Changing idenachain rights
chown -R $username:$username /home/$username/idena-go
#Continue as username
sudo -i -u $username bash << EOF
whoami
cd ~/idena-go
#
chmod +x idena-node
screen -d -m bash -c "/home/$username/idena-go/./idena-node --config=config.json"
echo Idena-node has been installed succesfully

until [ -s ~/idena-go/datadir/api.key ]
do
     sleep 5
done

until [ -s ~/idena-go/datadir/keystore/nodekey ]
do
     sleep 5
done

EOF

apikey=$( cat /home/$username/idena-go/datadir/api.key )
if [ "$nodeapikey" != 'n' ]; then
apikey="$nodeapikey"
( sudo -i -u $username echo "$apikey" >| /home/$username/idena-go/datadir/api.key )
fi
echo -e ${LBLUE}Your IDENA-node API key is: ${YELLOW}$apikey
prvkey=$( cat /home/$username/idena-go/datadir/keystore/nodekey )
if [ "$privatekey" != 'n' ]; then
prvkey="$privatekey"
( sudo -i -u $username echo "$prvkey" >| /home/$username/idena-go/datadir/keystore/nodekey ) 
fi
echo -e ${LBLUE}Your IDENA-node PRIVATE key is: ${YELLOW}$prvkey${NC}

#If yes changing prv and api keys

if [[ "$privatekey" = 'n' && "$nodeapikey" = 'n' ]]; then
while true; do
    read -p "$(echo -e ${LYELLOW}Would you like to add your own ${LGREEN}IDENA NODE PRIVATE KEY${LYELLOW} \(aka ${LRED}nodekey${LYELLOW}\)?$'\n'${LGREEN}Default path: ${LRED}../idena-go/datadir/keystore/nodekey ${LGREEN}[y/N]${NC} )" yn
    case $yn in
        [Yy]* ) killall screen; read -p "$(echo -e ${LYELLOW}Please enter your IDENA private key:${NC} )"; echo "$REPLY" >| /home/$username/idena-go/datadir/keystore/nodekey; break;;
        [Nn]* ) break;;
        * ) echo "Please answer yes or no.";;
    esac
done

while true; do
    read -p "$(echo -e ${LYELLOW}Would you like to add your own ${LGREEN}IDENA NODE API KEY${LYELLOW} \(aka ${LRED}api.key${LYELLOW}\)?$'\n'${LGREEN}Default path: ${LRED}../idena-go/datadir/api.key  ${LGREEN}[y/N]${NC} )" yn
    case $yn in
        [Yy]* ) killall screen > /dev/null 2>&1; read -p "$(echo -e ${LYELLOW}Please enter your IDENA node key: ${NC})"; echo "$REPLY" >| /home/$username/idena-go/datadir/api.key; break;;
        [Nn]* ) break;;
        * ) echo "Please answer yes or no.";;
    esac
done
fi
if [[ "$privatekey" = 'n' && "$nodeapikey" != 'n' ]]; then
while true; do
    read -p "$(echo -e ${LYELLOW}Would you like to add your own ${LGREEN}IDENA NODE PRIVATE KEY${LYELLOW} \(aka ${LRED}nodekey${LYELLOW}\)?$'\n'${LGREEN}Default path: ${LRED}../idena-go/datadir/keystore/nodekey ${LGREEN}[y/N]${NC} )" yn
    case $yn in
        [Yy]* ) killall screen > /dev/null 2>&1; read -p "$(echo -e ${LYELLOW}Please enter your IDENA private key:${NC} )"; echo "$REPLY" >| /home/$username/idena-go/datadir/keystore/nodekey; echo "$nodeapikey" >| /home/$username/idena-go/datadir/api.key; break;;
        [Nn]* ) break;;
        * ) echo "Please answer yes or no.";;
    esac
done
fi
if [[ "$nodeapikey" = 'n' && "$privatekey" != 'n' ]]; then
while true; do
    read -p "$(echo -e ${LYELLOW}Would you like to add your own ${LGREEN}IDENA NODE API KEY${LYELLOW} \(aka ${LRED}api.key${LYELLOW}\)?$'\n'${LGREEN}Default path: ${LRED}../idena-go/datadir/api.key  ${LGREEN}[y/N]${NC} )" yn
    case $yn in
        [Yy]* ) killall screen > /dev/null 2>&1; read -p "$(echo -e ${LYELLOW}Please enter your IDENA node key: ${NC})"; echo "$REPLY" >| /home/$username/idena-go/datadir/api.key; echo "$privatekey" >| /home/$username/idena-go/datadir/keystore/nodekey; break;;
        [Nn]* ) break;;
        * ) echo "Please answer yes or no.";;
    esac
done
fi
# kill screen before to create the idena daemon
killall screen
# creating idena daemon
sed -i "s/\$username/${username}/g" /home/$username/idena-go/idena.service
cp /home/$username/idena-go/idena.service /etc/systemd/system/idena_$username.service
systemctl daemon-reload
systemctl reset-failed
systemctl start idena_$username.service
systemctl enable idena_$username.service
#Checking for idena updates ones a day
cp idena_insp.sh /home/$username/idena-go/idena_insp_$username.sh
chown $username:$username /home/$username/idena-go/idena_insp_$username.sh
if [ "$idupdate" = 'n' ]; then
read -p $'\e[33mPlease insert \e[32mthe frequency \e[91min cron schedule expression format \e[33mwhen the script will be checking for updates. \n\e[31m\e[1mEmpty prompt \e[33mwill set the value to run it \e[32monce a day at 1AM:\e[0m ' idupdate
if [[ -z $idupdate ]]; then idupdate=$(echo "0 1 * * *") ; echo "Set as default $idupdate"; fi
fi    
echo "$idupdate $username /home/$username/idena-go/idena_insp_$username.sh" > /etc/cron.d/idena_update_$username
#crontab -l | grep -q "idena_insp_$username"  && echo 'entry exists' || (crontab -l 2>/dev/null; echo "$idupdate /home/$username/idena-go/idena_insp_$username.sh") | crontab -
# ufw configuration
SSHPORT=$(echo $SSH_CLIENT | awk '{print $3}')
ufw allow $SSHPORT
ufw allow "OpenSSH"
ipfsport=($(jq -r '.IpfsConf.IpfsPort' /home/$username/idena-go/config.json))
ufw allow ${ipfsport[0]} comment "IDENA Instance for user $username"
echo "y" | sudo ufw enable
#sudo ufw status
# Installation has been successfully completed
echo -e "${LRED}IDENA NODE HAS BEEN SUCCESSFULLY INSTALLED" 
echo -e "${LGREEN}FOR IDENA DONATIONS:${NC} 0xf041640788910fc89a211cd5bcbf518f4f14d831"
echo -e "${YELLOW}CONTACT AUTHOR:${NC} ltraveler@protonmail.com"
echo -e "${LBLUE}IDENA PERSONALIZED SHARED NODE SERVICE:${NC} https://t.me/ltrvlr"
exit
